### ISSUE TITLE
[Issue]: About retrieve_content and code execution in call_rag_chat

### ISSUE THREAD

**Issue by szwang-seedtech** [microsoft/autogen#1769]
### Describe the issue

I was following the agentchat_groupchat_RAG.ipynb https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat_RAG.ipynb

When I was using call_rag_chat, sometimes the agent will refuse to execute the code and terminate the conversation, but sometimes works fine. If I remove the retrieve_content part in call_rag_chat, then all the code will be executed by agents without any problem.

1. If I only listed user in register_for_execution, I will have the execution code problem occationally
2. if I listed user and other assistant agents in register_for_execution, I will have the error "openai.BadRequestError: Error code: 400 - {'error': {'message': "None is not of type 'object' - 'messages.2.function_call'", 'type': 'invalid_request_error', 'param': None, 'code': None}}" occationally

Does anyone have any clue, or is it a bug?

### Screenshots and logs
This is what I mean by "refuse to execute the code and terminate the conversation"
![Screenshot 2024-02-23 182733](https://github.com/microsoft/autogen/assets/43228022/1aba86c5-7caa-429a-b951-41c960b2b0f8)

**Comment by szwang-seedtech** [comment]
I am using --no-cache-dir sentence-transformers, will that be related?

**Comment by gagb** [comment]
@thinkall

**Comment by thinkall** [comment]
Hi @szwang-seedtech , could you share the whole notebook of your version so that I can check the outputs and reproduce the issue? Thanks.

**Comment by szwang-seedtech** [comment]
@thinkall 
I am just using the agentchat_groupchat_RAG.ipynb, and uses call_rag_chat. 
But instead of put both user and pm into register_for_execution,
1. if I only put user there, sometimes I will get no reply; 
2. if I put user and other agents, sometimes I will get the mentioned error "openai.BadRequestError: Error code: 400...", and it seems like the same with issue 1440? 

One more thing, I am using --no-cache-dir sentence transformer, do you think that could be related?

**Comment by szwang-seedtech** [comment]
@thinkall hello, attached is the code I am using.

import sys

__import__("pysqlite3")
sys.modules["sqlite3"] = sys.modules.pop("pysqlite3")

import chromadb
from typing_extensions import Annotated
from typing import Any, List
import autogen
import json
import re

from autogen import AssistantAgent, Cache
from autogen.agentchat.contrib.retrieve_user_proxy_agent import RetrieveUserProxyAgent


def termination_msg(x):
    print(f"SEED_MSG_START{json.dumps(x)}SEED_MSG_END")
    return isinstance(x, dict) and "TERMINATE" == str(x.get("content", ""))[-9:].upper()


seed_gpt_setting = None
seed_config_list = None
with open("./settings/SEED_GPT_SETTING.json") as f:
    seed_gpt_setting = json.load(f)

with open("./settings/SEED_CONFIG_LIST.json") as f:
    seed_config_list = json.load(f)

llm_groups = seed_gpt_setting.get("llm_groups")
user_proxy_agent = seed_gpt_setting.get("user_proxy_agent")
agent_list = seed_gpt_setting.get("agent_list")
problem = seed_gpt_setting.get("content")


def get_llm_config(model: List[str] or str):
    llm_list = autogen.config_list_from_json(
        env_or_file="SEED_CONFIG_LIST.json",
        file_location="./settings",
        filter_dict={
            "model": type(model) is list and model or [model],
        },
    )
    llm_config = {
        "config_list": llm_list,
        "timeout": 60,
        "cache_seed": user_proxy_agent.get("cache_seed"),
        "temperature": 0,
    }
    return llm_list, llm_config


def get_user_proxy_agent(name: str, system_message: str, default_auto_reply: str):
    user = autogen.UserProxyAgent(
        name=name,
        description="seed",
        is_termination_msg=termination_msg,
        human_input_mode="NEVER",
        code_execution_config=False,  # we don't want to execute code in this case.
        system_message=system_message,
        default_auto_reply=default_auto_reply,
    )

    return user


def get_docs_path(docs_path: str):
    ret = [
        "https://raw.githubusercontent.com/microsoft/FLAML/main/website/docs/Examples/Integrate%20-%20Spark.md"
    ]
    if docs_path:
        filename = re.split(r"[\\/]", docs_path)[-1]
        ret.append(f"/app/docs/{filename}")
    return ret


def get_assistant_proxy_agent(
    assistant_model: str, assistant_name: str, docs_path: str
):
    rag_llm, _ = get_llm_config(assistant_model)
    rag_aid = RetrieveUserProxyAgent(
        name=assistant_name,
        is_termination_msg=termination_msg,
        system_message="Assistant who has extra content retrieval power for solving difficult problems.",
        human_input_mode="NEVER",
        max_consecutive_auto_reply=3,
        retrieve_config={
            "docs_path": get_docs_path(docs_path),
            "chunk_token_size": 1000,
            "model": rag_llm[0]["model"],
            "client": chromadb.PersistentClient(path="/tmp/chromadb"),
            "get_or_create": True,
        },
        code_execution_config=False,  # we don't want to execute code in this case.
    )

    return rag_aid


def get_agent(agent_config_list: Any):
    agent_ret = []

    for agent_config in agent_config_list:
        model_list = None
        for group in llm_groups:
            if group.get("llm_list_name") == agent_config.get("agent_model_list"):
                model_list = group.get("model_list")

        agent_llm_list, agent_llm_config = get_llm_config(model_list)

        agent_item = AssistantAgent(
            name=agent_config.get("agent_name"),
            description="seed",
            is_termination_msg=termination_msg,
            system_message=agent_config.get("agent_system_message"),
            llm_config=agent_llm_config,
            code_execution_config=(
                {
                    "last_n_messages": "auto",
                    "work_dir": "./seed-work-dir",
                    "use_docker": False,
                }
                if agent_config.get("code_execution")
                else False
            ),
        )

        agent_ret.append(
            {
                "ret": agent_item,
                "llm_list": agent_llm_list,
            }
        )

    return agent_ret


def register_all_llm():
    custom_llm_list = {}
    custom_llm_config = {}
    manager_llm_config = None

    for llm_group in llm_groups:
        list_name = f"${llm_group.get('llm_list_name')}_list"
        group_name = f"${llm_group.get('llm_list_name')}_config"
        llm_list, llm_config = get_llm_config(llm_group.get("model_list"))

        custom_llm_list[list_name] = llm_list
        custom_llm_config[group_name] = llm_config

    _, manager_llm_config = get_llm_config(user_proxy_agent.get("manager_model"))

    return custom_llm_config, manager_llm_config


def main():
    _, manager_llm_config = register_all_llm()
    user = get_user_proxy_agent(
        name=user_proxy_agent.get("user_name"),
        system_message=user_proxy_agent.get("system_message"),
        default_auto_reply=user_proxy_agent.get("user_reply_message"),
    )
    aid = get_assistant_proxy_agent(
        assistant_model=user_proxy_agent.get("assistant_model"),
        assistant_name=user_proxy_agent.get("assistant_name"),
        docs_path=user_proxy_agent.get("docs_path"),
    )
    agent_ret = get_agent(agent_list)
    agent_ret_list = [agent.get("ret") for agent in agent_ret]

    def _reset_agents():
        aid.reset()
        user.reset()
        for agent in agent_ret_list:
            agent.reset()

    _reset_agents()

    def retrieve_content(
        message: Annotated[
            str,
            "Refined message which keeps the original meaning and can be used to retrieve content for code generation and question answering.",
        ],
        n_results: Annotated[int, "number of results"] = 3,
    ) -> str:
        aid.n_results = n_results
        update_context_case1, update_context_case2 = aid._check_update_context(message)
        if (update_context_case1 or update_context_case2) and aid.update_context:
            aid.problem = message if not hasattr(aid, "problem") else aid.problem
            _, ret_msg = aid._generate_retrieve_user_reply(message)
        else:
            ret_msg = aid.generate_init_message(message, n_results=n_results)
        return ret_msg if ret_msg else message

    aid.human_input_mode = "NEVER"

    for caller in agent_ret_list:
        d_retrieve_content = caller.register_for_llm(
            description="retrieve content for code generation and question answering.",
            api_style="function",
        )(retrieve_content)

    for executor in [user]:
        executor.register_for_execution()(d_retrieve_content)

    groupchat = autogen.GroupChat(
        agents=[user] + agent_ret_list,
        messages=[],
        max_round=user_proxy_agent.get("max_round"),
        speaker_selection_method="auto",
        allow_repeat_speaker=False,
    )

    manager = autogen.GroupChatManager(
        groupchat=groupchat, llm_config=manager_llm_config
    )

    with Cache.disk(
        cache_path_root="./settings/.cache",
        cache_seed=user_proxy_agent.get("cache_seed"),
    ) as cache:
        user.initiate_chat(manager, message=problem, cache=cache)


main()

**Comment by thinkall** [comment]
Thanks for providing the info, but the code can't run.
```
with open("./settings/SEED_GPT_SETTING.json") as f:
    seed_gpt_setting = json.load(f)

with open("./settings/SEED_CONFIG_LIST.json") as f:
    seed_config_list = json.load(f)
```
These two files are needed.

For the code execution issue, do you mean running the code generated by LLM, or call the rag function?
For the #2, I think it's function call error.

**Comment by szwang-seedtech** [comment]
@thinkall SEED_GPT_SETTING is the SETTING file I attached. SEED_CONFIG_LIST is just a file with config_list, so that main.py can load models from there.

And yes, the code execution issue is the code generated by LLM.

Do you have any idea how I can fix the function call error? I noticed the same error in issue 1440, does it get fix in the newly released 0.2.15?

**Comment by szwang-seedtech** [comment]
For more details on "code execution issue", when I ask agents to write a code, for example "write a bubble sorting sample code", that is unrelevant to the doc I provided to the rag, agents will be busy calling retrieve_content, and only send retrieve_content with empty messages to each other and then terminate the conversation. Looks like in a retrieve_content loop.

**Comment by thinkall** [comment]
> For more details on "code execution issue", when I ask agents to write a code, for example "write a bubble sorting sample code", that is unrelevant to the doc I provided to the rag, agents will be busy calling retrieve_content, and only send retrieve_content with empty messages to each other and then terminate the conversation. Looks like in a retrieve_content loop.

In your code, you're using `speaker_selection_method="auto"`, I think this is the root cause. The speaker selection result from the LLM doesn't work well in your case. I suggest you try `round_robin`.

**Comment by thinkall** [comment]
> @thinkall SEED_GPT_SETTING is the SETTING file I attached. SEED_CONFIG_LIST is just a file with config_list, so that main.py can load models from there.
> 
> And yes, the code execution issue is the code generated by LLM.
> 
> Do you have any idea how I can fix the function call error? I noticed the same error in issue 1440, does it get fix in the newly released 0.2.15?

The fix for 1440 is just updating the usage in the notebook. Your code is already using the latest APIs for function calling.

**Comment by szwang-seedtech** [comment]
Thank you! Your replies give me some inspiration, and I am testing it.
